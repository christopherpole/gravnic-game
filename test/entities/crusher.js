import {
  MOVE_UP,
  MOVE_DOWN,
  MOVE_RIGHT,
  ENTITIES,
  changeGravityDirection,
  levelIsComplete,
} from '../../src/';

describe('Crusher entity', () => {
  it('Should crush a single movable entity', () => {
    const gameState = [
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 2 },
          movableEntity: { entityId: ENTITIES.BLOCK.id, color: '#ff0000', id: 1 },
        },
      ],
      [{ staticEntity: { entityId: ENTITIES.FLOOR.id, id: 3 }, movableEntity: null }],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 5 },
          movableEntity: { entityId: ENTITIES.CRUSHER.id, id: 4 },
        },
      ],
    ];

    const gameStates = changeGravityDirection(gameState, MOVE_UP);

    expect(gameStates).toMatchSnapshot();
    expect(levelIsComplete(gameStates[gameStates.length - 1])).toBe(true);
  });

  it('Should crush multiple movable entities', () => {
    const gameState = [
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 2 },
          movableEntity: { entityId: ENTITIES.GLASS.id, id: 1 },
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 4 },
          movableEntity: { entityId: ENTITIES.SMART_BOMB.id, id: 3 },
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 6 },
          movableEntity: { entityId: ENTITIES.CRATE.id, id: 5 },
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 8 },
          movableEntity: { entityId: ENTITIES.BLOCK.id, color: '#ff0000', id: 7 },
        },
      ],
      [{ staticEntity: { entityId: ENTITIES.FLOOR.id, id: 9 }, movableEntity: null }],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 11 },
          movableEntity: { entityId: ENTITIES.CRUSHER.id, id: 10 },
        },
      ],
    ];

    const gameStates = changeGravityDirection(gameState, MOVE_UP);

    expect(gameStates).toMatchSnapshot();
    expect(levelIsComplete(gameStates[gameStates.length - 1])).toBe(true);
  });

  it('Should not crush static entities', () => {
    const gameState = [
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 2 },
          movableEntity: { entityId: ENTITIES.BLOCK.id, color: '#800080', id: 1 },
        },
      ],
      [{ staticEntity: { entityId: ENTITIES.STICKY_SPOT.id, id: 3 }, movableEntity: null }],
      [
        {
          staticEntity: { entityId: ENTITIES.GRAVITY_CHANGER.id, direction: MOVE_UP, id: 4 },
          movableEntity: null,
        },
      ],
      [
        {
          staticEntity: {
            entityId: ENTITIES.COLOR_CHANGER.id,
            targetEntity: { entityId: ENTITIES.BLOCK.id, color: '#ff0000' },
            id: 5,
          },
          movableEntity: null,
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 7 },
          movableEntity: { entityId: ENTITIES.CRUSHER.id, id: 6 },
        },
      ],
    ];

    const gameStates = changeGravityDirection(gameState, MOVE_UP);

    expect(gameStates).toMatchSnapshot();
    expect(levelIsComplete(gameStates[gameStates.length - 1])).toBe(false);
  });

  it('Should crush movable entities that move in to it', () => {
    const gameState = [
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 2 },
          movableEntity: { entityId: ENTITIES.BLOCK.id, color: '#0000FF', id: 1 },
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 4 },
          movableEntity: { entityId: ENTITIES.BLOCK.id, color: '#FFE4E1', id: 3 },
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 6 },
          movableEntity: { entityId: ENTITIES.GLASS.id, id: 5 },
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 8 },
          movableEntity: { entityId: ENTITIES.CRUSHER.id, id: 7 },
        },
      ],
    ];

    const gameStates = changeGravityDirection(gameState, MOVE_DOWN);

    expect(gameStates).toMatchSnapshot();
    expect(levelIsComplete(gameStates[gameStates.length - 1])).toBe(true);
  });

  it('Should not crush entities while stuck', () => {
    const gameState = [
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 2 },
          movableEntity: { entityId: ENTITIES.CRUSHER.id, id: 1 },
        },
        { staticEntity: { entityId: ENTITIES.STICKY_SPOT.id, id: 3 }, movableEntity: null },
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 5 },
          movableEntity: { entityId: ENTITIES.BLOCK.id, color: '#ff0000', id: 4 },
        },
      ],
    ];

    const gameStates = changeGravityDirection(gameState, MOVE_RIGHT);

    expect(gameStates).toMatchSnapshot();
    expect(levelIsComplete(gameStates[gameStates.length - 1])).toBe(false);
  });

  it('Should not let stuck entities fall into it', () => {
    const gameState = [
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 2 },
          movableEntity: { entityId: ENTITIES.CRATE.id, id: 1 },
        },
        { staticEntity: { entityId: ENTITIES.FLOOR.id, id: 3 }, movableEntity: null },
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 5 },
          movableEntity: { entityId: ENTITIES.CRUSHER.id, id: 4 },
        },
      ],
      [
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 7 },
          movableEntity: { entityId: ENTITIES.BLOCK.id, color: '#ff0000', id: 6 },
        },
        { staticEntity: { entityId: ENTITIES.STICKY_SPOT.id, id: 8 }, movableEntity: null },
        {
          staticEntity: { entityId: ENTITIES.FLOOR.id, id: 10 },
          movableEntity: { entityId: ENTITIES.CRUSHER.id, id: 9 },
        },
      ],
    ];

    const gameStates = changeGravityDirection(gameState, MOVE_RIGHT);

    expect(gameStates).toMatchSnapshot();
    expect(levelIsComplete(gameStates[gameStates.length - 1])).toBe(false);
  });
});
