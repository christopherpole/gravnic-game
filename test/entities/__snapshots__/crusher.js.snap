// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crusher entity Should crush a single movable entity 1`] = `
Array [
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "color": "#ff0000",
          "entityId": "BLOCK",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 4,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 3,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 5,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 4,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 3,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 5,
        },
      },
    ],
  ],
]
`;

exports[`Crusher entity Should crush movable entities that move in to it 1`] = `
Array [
  Array [
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "color": "#0000FF",
          "entityId": "BLOCK",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "color": "#FFE4E1",
          "entityId": "BLOCK",
          "id": 3,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 7,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "color": "#0000FF",
          "entityId": "BLOCK",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 7,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 7,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
  ],
]
`;

exports[`Crusher entity Should crush multiple movable entities 1`] = `
Array [
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "GLASS",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "SMART_BOMB",
          "id": 3,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRATE",
          "id": 5,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "color": "#ff0000",
          "entityId": "BLOCK",
          "id": 7,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 10,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 9,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 11,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "GLASS",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "SMART_BOMB",
          "id": 3,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRATE",
          "id": 5,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 10,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 9,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 11,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "GLASS",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "SMART_BOMB",
          "id": 3,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 10,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 9,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 11,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "GLASS",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 10,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 9,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 11,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 10,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 6,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 8,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 9,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 11,
        },
      },
    ],
  ],
]
`;

exports[`Crusher entity Should not crush entities while stuck 1`] = `
Array [
  Array [
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 1,
          "stuck": true,
        },
        "staticEntity": Object {
          "entityId": "STICKY_SPOT",
          "id": 3,
        },
      },
      Object {
        "movableEntity": Object {
          "color": "#ff0000",
          "entityId": "BLOCK",
          "id": 4,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 5,
        },
      },
    ],
  ],
]
`;

exports[`Crusher entity Should not crush static entities 1`] = `
Array [
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "color": "#800080",
          "entityId": "BLOCK",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "STICKY_SPOT",
          "id": 3,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "direction": "MOVE_UP",
          "entityId": "GRAVITY_CHANGER",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 6,
        },
        "staticEntity": Object {
          "entityId": "COLOR_CHANGER",
          "id": 5,
          "targetEntity": Object {
            "color": "#ff0000",
            "entityId": "BLOCK",
          },
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 7,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "color": "#800080",
          "entityId": "BLOCK",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "STICKY_SPOT",
          "id": 3,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 6,
        },
        "staticEntity": Object {
          "direction": "MOVE_UP",
          "entityId": "GRAVITY_CHANGER",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "COLOR_CHANGER",
          "id": 5,
          "targetEntity": Object {
            "color": "#ff0000",
            "entityId": "BLOCK",
          },
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 7,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "color": "#800080",
          "entityId": "BLOCK",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 6,
          "stuck": true,
        },
        "staticEntity": Object {
          "entityId": "STICKY_SPOT",
          "id": 3,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "direction": "MOVE_UP",
          "entityId": "GRAVITY_CHANGER",
          "id": 4,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "COLOR_CHANGER",
          "id": 5,
          "targetEntity": Object {
            "color": "#ff0000",
            "entityId": "BLOCK",
          },
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 7,
        },
      },
    ],
  ],
]
`;

exports[`Crusher entity Should not let stuck entities fall into it 1`] = `
Array [
  Array [
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
      Object {
        "movableEntity": Object {
          "entityId": "CRATE",
          "id": 1,
          "moved": true,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 3,
        },
      },
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 4,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 5,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 7,
        },
      },
      Object {
        "movableEntity": Object {
          "color": "#ff0000",
          "entityId": "BLOCK",
          "id": 6,
          "stuck": true,
        },
        "staticEntity": Object {
          "entityId": "STICKY_SPOT",
          "id": 8,
        },
      },
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 9,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 10,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
      Object {
        "movableEntity": Object {
          "entityId": "CRATE",
          "id": 1,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 3,
        },
      },
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 4,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 5,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 7,
        },
      },
      Object {
        "movableEntity": Object {
          "color": "#ff0000",
          "entityId": "BLOCK",
          "id": 6,
          "stuck": true,
        },
        "staticEntity": Object {
          "entityId": "STICKY_SPOT",
          "id": 8,
        },
      },
      Object {
        "movableEntity": Object {
          "entityId": "CRUSHER",
          "id": 9,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 10,
        },
      },
    ],
  ],
]
`;
